### ISSUE TITLE
Feature Request: More control over allowed authentication methods

### ISSUE THREAD

**Issue by colstrom** [hirura/hrr_rb_ssh#10]
There's a few items tangled together in this request that are all related.

`Authentication::Method.list_preferred` relies on a `PREFERENCE` constant in each class. While this can be overridden, doing so spews warnings about redefining constants (as it should!). It would be preferable for this to be configurable.

`Authentication#send_userauth_failure` has `:'partial success'` fixed as `false`. Being able to report partial success to the client would be convenient for multi-step authentication flows.
 
It would be helpful to be able to alter the allowed authentication methods on a per-session basis. This would allow for user-specific authentication flows (such as requiring 2FA if the user has configured it).

I'm currently resolving that last one with the following in my authenticator class (not suggesting this as a proper solution, just for reference):

```ruby
    private

    def authentications
      ::ObjectSpace.each_object.select do |object|
        ::HrrRbSsh::Authentication === object
      end
    end

    def authentication
      authentications.find do |authentication|
        context.variables.equal?(::Kernel
                                   .instance_method(:instance_variable_get)
                                   .bind(authentication)
                                   .call(:@variables))
      end
    end

    def can_continue(authentication_methods = ::HrrRbSsh::Authentication::Method.list_preferred, partial_success = false)
      authentication.define_singleton_method(:send_userauth_failure) do
        {
         :"message number"                    => ::HrrRbSsh::Message::SSH_MSG_USERAUTH_FAILURE::VALUE,
         :"authentications that can continue" => ::Kernel.Array(authentication_methods),
         :"partial success"                   => partial_success,
        }
          .then { |message| ::HrrRbSsh::Message::SSH_MSG_USERAUTH_FAILURE.encode(message) }
          .then { |payload| instance_variable_get(:@transport).send(payload) }
      end
```

This works, but there should probably be a better way. ;p

**Comment by hirura** [comment]
I have tried implementing the multi-step authentication. Could you take a look at PR #11 ?

**Comment by colstrom** [comment]
Reviewing #11, it addresses the concerns raised, though there's a few things that feel a bit... off? Maybe these are minor quibbles, so I'll just lay them out here and you can judge them accordingly:

The idea that a full set of partial successes implies a successful login seems *technically* valid, but feels counterintuitive. If "partial success" means "this check passed, but we need something else", then eventually (in my head) something should have to say "okay, we're good now" and return success, not partial success, for a login to be accepted.

This approach only allows *removing* authentication methods. Given that the authentication flow generally (in a typical client) starts with a `none` attempt, being able to *add* types would allow starting with a minimal set of auth methods, and adding the ones that are possible for a given user. As an example: `publickey` always allowed, `password` allowed if the user has set a password, and if they have both password and 2FA configured, then `password` is replaced with `keyboard-interactive`. Or something like that.

Other than that, it looks like a solid improvement!

**Comment by hirura** [comment]
Thanks for the comments. I understand that and the PR may have further improvements.
Then I will implement a bit more and define a sort of strategy of implementation of authentication handling.

1. Simple, existing authentication

Each authenticator returns `true` (or `HrrRbSsh::Authentication::SUCCESS`) or `false` (or `HrrRbSsh::Authentication::FAILURE`). When it is true, the user is accepted. When it is false, the user is not accepted and a subsequent authenticator is called.

2. Multi-step authentication by pre-defined preferred authentication methods and partial success (PR #11)

In this strategy, an implementer doesn't think about what authentication method requested in what order. Just expecting that all the authentications that is listed in a preferred authentication methods are accepted and their authenticator returns `HrrRbSsh::Authentication::PARTIAL_SUCCESS`.

3. More flexible authentication (PR #xx)

A `context` variable in authenticator gives additional access to remaining authentication methods that can continue. In this strategy, an implementer is able to control the order of authentication methods and to control which authentication methods are used for the user.

The below pseudo-code is an example. It is expected that any user must be verified by publickey and then another authentication is requested for the user accordingly.

```ruby
options['preferred_authentication_methods'] = ['none']

auth_none = HrrRbSsh::Authentication::Authenticator.new{ |context|
  context.remaining_authentication_methods.push 'publickey'
  HrrRbSsh::Authentication::PARTIAL_SUCCESS
}

auth_publickey = HrrRbSsh::Authentication::Authenticator.new{ |context|
  if some_verification(context)
    case context.username
    when 'user1'
      context.remaining_authentiation_methods.push 'keyboard-interactive'
    else
      context.remaining_authentiation_methods.push 'password'
    end
    HrrRbSsh::Authentication::PARTIAL_SUCCESS
  else
    false
  end
}

auth_keyboard_interactive = HrrRbSsh::Authentication::Authenticator.new{ |context|
  if some_verification(context)
    true # or HrrRbSsh::Authentication::PARTIAL_SUCCESS; both will accept the user because remaining authentication method is only 'keyboard-interactive' in this case
  else
    false
  end
}
```

**Comment by hirura** [comment]
I have merged PR #11 and #12, and released [0.3.0.pre3](https://rubygems.org/gems/hrr_rb_ssh/versions/0.3.0.pre3). Please try it !

**Comment by colstrom** [comment]
Tested with `0.3.0.pre3`, and it works like a charm, thanks! I was able to add/remove authentication methods easily, and handle partial success as described.

Your description of partial success (specifically in the case where the *order* of authentication methods is unclear) has convinced me that it makes sense to do it that way. I was overlooking how helpful it could be in that scenario.

As a side note, with this new functionality, it's also possible to force a given order of authentication methods by only presenting the ones that are acceptable next steps.

If I may bloat the scope of this request just a **bit** more...

It would be nice to be able to determine the status of the authentication methods available for a session. So being able to query if say... password authentication is :success, :failure, or :partial_success (or some status indicating "not attempted yet" like :pending).

This would allow things like gating access to features/subsystems based on authentication method. Maybe the only thing a user can do without a password is reset their password, or perhaps a dangerous operation like "delete everything" requires at least two forms of authentication.

**Comment by hirura** [comment]
Thanks for the testing. That sounds great !

Regard to such an access controll, I think, it's not the library's matter. Thanks to your earlier issue #7, it is now possible to share context.variables between authenticators and channel request handlers, so it can be implemented with it.

**Comment by colstrom** [comment]
Oh, the **access control** stuff definitely seems out of scope. That's very specific to what I'm doing, and was only intended as an example of where it would be helpful to be able to *check which authentication methods were attempted, and the status of each*.

This is completely *possible* now, and I have been using `context.variables` to do so, so we're in agreement there. It seemed to me like something that made sense in this library, but perhaps it's less generally useful than I suspected.

Thanks for considering it, regardless!

**Comment by hirura** [comment]
That's good to hear that your use-case is covered by the current library implementation. I agree with you that providing every authentication results in a channel request handler may be useful for some situation, but I don't have a good way to do that. So let us postpone the item indefinitely, so far. When we find a good use-case or a way to give the access, then I'll try to implement that.

Now I think this issue here can be closed. Let me close the issue when I release version 0.3.0. Honestly I'm not sure when I can release the version and what the version updates...

Thanks,

**Comment by colstrom** [comment]
That sounds like a reasonable plan.

As far as what was updated along the way, I skimmed over `git log v0.2.2..v0.3.0.pre3`, and it looks like the highlights were:
  - IO is passed to Server#start, rather than Server.new.
  - Transport option for custom local version.
  - Context#close_session was added.
  - Can share variables between authenticator and request handler contexts.
  - Better support for multi-step authentication flows.
    - Support for Partial Success
    - Able to change allowed authentication methods on a per-session basis.

I appreciate the work that's gone into these (and this library in general). You've given me a nice foundation to build on!

**Comment by hirura** [comment]
I have released v0.3.0, so close the issue.

Thanks for the summarizing updates. Now v0.3.0 updates are:
  - IO is passed to Server#start, rather than Server.new.
  - Transport option for custom local version.
  - Context#close_session was added.
  - Can share variables between authenticator and request handler contexts.
  - Better support for multi-step authentication flows.
    - Support for Partial Success
    - Able to change allowed authentication methods on a per-session basis.
  - Client mode support (Experimental)
  - Logger to output process and thread ID
